proposals/0000-namespaces.md

```diff
@@ -0,0 +1,778 @@
+    different arity or arguments.
+
+3.  Renaming things. This is the most common solution. Pre- or Suffixing namespace identifies to solutions; 
+    e.g. `prettyExpr`, `prettyType`, `prettyVar`, `MkT`, `runState`, `applyEndo`, and so on. 
+    Besides (subjectivly) being akward to use and read, it has multiple drawbacks.
+    
+    a) It prompotes unqualified import of modules, e.g., `State.runState` seems redundant.
+       This is turn will pull more items into scope making it harder to present good completion
+       options for the user of an IDE.
+    
+    b) With suffix notation is hard to discover the abilities of a types using an IDE, e.g., 
+       typing `pretty` will yield `{prettyExpr, prettyType, prettyVar}`, but typing 
+       `Expr:` could yield `{Expr:pretty, Expr:evaluate, Expr:normalize, ...}`.
+    
+    c) The trick does not work for operators.
+
+The rest of the motivation is a list of examples mirroring [#283](https://github.com/ghc-proposals/ghc-proposals/pull/283)
+
+### Example 1: Compact code
+
+If we want associate the same named function with two data structures, we'll have 
+to create two files, create a type class, or prefix the names of the function.
+
+With namespaces you can write this:
+
+```haskell
+module Vec where
+
+namespace V2 where
+  data V2 = Mk { x :: Int, y :: Int }
+  fromV2 :: V2 -> V2
+
+namespace V3 where
+  data V3 = Mk { x :: Int, y :: Int, z :: Int }
+  fromV2 :: V2 -> Int -> V3
+
+-- Creates the vector (V3:Mk 0 1 1) 
+vec_0_1_1 :: V3:V3
+vec_0_1_1 = V3:fromV2 (V2:fromV2 (V2:Mk 0 1)) 1
+```
+
+This example creates two namespaces to contain the V2 and V3 data structures, 
+their field accessors, constructors, and their
+associated functions `fromV2`. 
+We can access the inners of `V2` and `V3` using the namespace operator, `:`, 
+this means that `V2:fromV2` points to the `fromV2` function in `V2`.
+
+This illustrates a common problem that you cannot easily fix with type classes, multiple functions
+with the same name, but different arguments. I that case you have to use adhoc-namespaces:
+
+```haskell
+module Vec where
+
+data V2 = MkV2 { v3x :: Int, v3y :: Int }
+fromV2V2 :: V2 -> V2
+
+data V3 = MkV3 { v3x :: Int, v3y :: Int, v3z :: Int }
+fromV2V3 :: V2 -> Int -> V3
+
+vec_0_1_1 :: V3
+vec_0_1_1 = fromV2V3 (fromV2V2 (MkV2 0 1)) 1
+```
+
+Or split it over multiple files:
+
+```haskell
+-- src/Vec/V2.hs
+module Vec.V2 where
+data V2 = Mk { x :: Int, y :: Int }
+fromV2 :: V2 -> V2
+```
+
+```haskell
+-- src/Vec/V3.hs
+module Vec.V3 where
+import Vec.V2 qualified as V2
+data V3 = Mk { x :: Int, y :: Int, z :: Int }
+fromV2 :: V2.V2 -> Int -> V3
+```
+
+```haskell
+-- src/Vec.hs
+module Vec where
+import Vec.V2 qualified as V2
+import Vec.V3 qualified as V3
+
+vec_0_1_1 :: V3.V3
+vec_0_1_1 = V3.fromV2 (V2.fromV2 (V2.Mk 0 1)) 1
+```
+
+However, in the last case you can't write a function `fromV3` in `Vec.V2` as it would 
+create a cyclic dependency.
```

Comment:
One technical clarification: GHC actually supports cyclic module dependencies so you **can** write `fromV3` in `Vec.V2`.

* https://downloads.haskell.org/ghc/latest/docs/users_guide/separate_compilation.html#how-to-compile-mutually-recursive-modules

Reply:
Good catch! I'll note that it is possible using a `hs-boot` file.